### PR TITLE
Fix getsockopt syscallbuf signal restart

### PR DESCRIPTION
**Pull request by yuyichao**
On the current asahi kernel and on the test machine (but not on the rb5 kernel
or the ArchLinux x86_64 kernel), somehow the `getsockopt` in `netfilter` test
takes 10-20 times longer to finish and may also return `EFAULT` when interrupted
by a signal. This broke restart of the syscall, which expect `ERESTART*` instead.

I strongly suspect both of these are caused by the `CONFIG_BPFILTER_UMH` kernel option
and in particular the `int bpfilter_send_req(struct mbox_request *req)`
in `bpfilter_kern.c` that talks to the user mode helper.
Assuming I understood how this works, it seems that this function
always returns `EFAULT` on error no matter what the error was.

While this can happen for user code regardless of rr,
it would be nice to reduce the effect of signals caused by rr.
For now, this happens in the test only for `netfilter` test in the untraced call
and can be fixed with a single manual restart of the syscall.
I believe restarting the syscall returning EFAULT should be safe here
since it should either fix the spurious failure or just return EFAULT again.

This could in principle happen out of syscall buf as well,
e.g. due to ticks interrupt but we can fix that later when it comes up.
(In such case at least we should have a orig_x0 to work with and don't have to
pass an extra one in as what we need for the untraced syscall).

--------------

This is still a hack but I feel like I understand enough of it to justify the hack... I tested the kernel behavior with the following program modified from `netfilter.c`

```c
#include "util.h"
#include <linux/netfilter_ipv4/ip_tables.h>

#include "nsutils.h"

#include <signal.h>

static void handler(int sig)
{
  (void)sig;
}

int main(void) {
  if (-1 == try_setup_ns(CLONE_NEWNET)) {
    atomic_puts("EXIT-SUCCESS");
    return 0;
  }

  signal(SIGALRM, handler);
  ualarm(1, 1);

  int sock_fd;

  sock_fd = socket(AF_INET, SOCK_RAW, IPPROTO_RAW);
  test_assert(sock_fd >= 0);

  for (int i = 0; i < 1000000000; i++) {
    struct ipt_getinfo info;
    memset(&info, 0, sizeof(info));
    strcpy(info.name, "nat");
    uint32_t getinfo_size = sizeof(info);
    errno = 0;
    int ret = getsockopt(sock_fd, SOL_IP, IPT_SO_GET_INFO, &info, &getinfo_size);
    if (ret < 0) {
      atomic_printf("%d\n", errno);
      return 1;
    }
  }
  atomic_puts("Finish");
  return 0;
}
```

On the two machines that are not affected, this will finish in a few minutes whereas on the two machines that are affected, this returns with a 14 (`EFAULT`) printed almost immediately.

### CODE REVIEW

**Comment by yuyichao** [comment]
I removed the "aarch64" from the commit title since I think in principle this could happen on x86 or at least the register difference shouldn't be an issue. The register difference makes it harder to restart the call manually but I assume it's the return value of the syscall that the kernel cares about, no mater where it is stored. OTOH, this code isn't particularly new... I feel like someone got to have run it on a x86 kernel with the correct configuration by now.....

I don't currently have a x86 kernel to test it but I can probably build one to test it later. That might also confirm the hypothesis on which option affects it if I do see it happen...

File: src/record_signal.cc [hunk 1]
```diff
@@ -411,6 +411,29 @@ static void handle_desched_event(RecordTask* t) {
    * That may be a kernel bug, but we handle it by just
    * continuing until we we continue past the arm-desched
    * syscall *and* stop seeing signals. */
+
+  const auto untraced_record_only_exit =
+    uintptr_t(RR_PAGE_SYSCALL_UNTRACED_RECORDING_ONLY) + 4;
```
Comment (rocallahan) [inline src/record_signal.cc:416]: This +4 is ARM-specific right? Fix this to be correct for all architectures?

File: src/record_signal.cc [hunk 2]
```diff
@@ -411,6 +411,29 @@ static void handle_desched_event(RecordTask* t) {
...
+  if (t->arch() == aarch64 && t->ip() == remote_code_ptr(untraced_record_only_exit) &&
```
Comment (rocallahan) [inline src/record_signal.cc:417]: Remove the architecture check?

**Review by rocallahan — CHANGES_REQUESTED** [review]

**Comment by rocallahan** [comment]
Could you report this kernel bug upstream?

File: src/record_signal.cc (hunk 1 above)
Comment (yuyichao) [reply src/record_signal.cc:416]: Ah you are right…. Apparently I’m too used to fixed instruction with now…

File: src/record_signal.cc (hunk 2 above)
Comment (yuyichao) [reply src/record_signal.cc:417]: Right. I’ll test on my laptop with a different kernel first.

File: src/record_signal.cc (hunk 1 above)
Comment (rocallahan) [reply src/record_signal.cc:416]: On x86 we usually can't have nice things ... but at least the syscall instructions all have the same size!

Comment (rocallahan) [reply src/record_signal.cc:416]: actually you can probably have variable length syscall instructions with prefixes. I guess that would still "work".

**Comment by yuyichao** [comment]
> Could you report this kernel bug upstream?

https://bugzilla.kernel.org/ Right?
Never had good luck before but I’ll try if I can confirm…

**Comment by yuyichao** [comment]
OK this is indeed reproducible on x86 using a stock arch linux kernel with `CONFIG_BPFILTER_UMH=m`. I had to adjust the sigalarm firing interval a little bit from 1us to 10us (and I actually also need to do that on my M1 after a reboot...) but otherwise both the test program and the netfilter test fails quite reliably on x86/x86_64 with the expected error. I'll report this upstream and fix the pr to handle x86.

**Comment by yuyichao** [comment]
Bug report upstream https://bugzilla.kernel.org/show_bug.cgi?id=216193 I'll amend in the commit message if I remember....

**Comment by rocallahan** [comment]
> Never had good luck before

One time I reported a bug there and someone fixed it a year later. But yes, it does require good luck :-).

**Comment by yuyichao** [comment]
Updated to work on both x86 and aarch64. Not as simple as just removing the arch check of course since the restart sequence is quite different...

**Comment by yuyichao** [comment]
As an aside, this was the actual failure that I saw that led to https://github.com/rr-debugger/rr/pull/3272 ... The premature failure with syscallbuf due to #3272 was initially masking this issue most of the time. So maybe someone did run the rr test suite on an affected kernel before but they may not have noticed this issue due to the other bug. I'm less surprised if no one has hit this issue in the wild outside of rr test suite, especially since no one has notice #3272 before....

**Pull request merged**